README.md
# godot-logger

[![Crates.io](https://img.shields.io/crates/v/godot-logger)](https://crates.io/crates/godot-logger)
[![docs.rs](https://img.shields.io/docsrs/godot-logger)](https://docs.rs/godot-logger)
[![GitHub branch checks state](https://img.shields.io/github/checks-status/jdno/godot-logger/main)](https://github.com/jdno/godot-logger/actions)

_A logger for [godot-rust] projects_

`godot-logger` is an easy-to-use logger for [godot-rust] projects. It prints
logs to Godot's output console and supports module-specific log levels.

## Usage

Start by adding [`godot-logger`] and [`log`] as dependencies to your project's
`Cargo.toml`.

```toml
[dependencies]
godot-logger = "1.1.0"
log = "0.4"
```

Then configure and initialize the logger in the `init` method that is passed to
`godot_init!`.

```rust
use gdnative::prelude::*;
use godot_logger::GodotLogger;
use log::{Level, LevelFilter};

fn init(handle: InitHandle) {
    if let Err(error) = GodotLogger::builder()
        .default_log_level(Level::Info)
        .add_filter("godot_logger", LevelFilter::Debug)
        .init()
    {
        godot_warn!("{}", error.to_string());
    }

    log::debug!("Initialized the logger");
}

godot_init!(init);
```

The following will be printed in the _Output_ console inside Godot:

```text
2021-09-25 19:29:25 DEBUG godot_logger Initialized the logger
```

## License

Licensed under either of

- Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or <http://www.apache.org/licenses/LICENSE-2.0>)
- MIT license ([LICENSE-MIT](LICENSE-MIT) or <http://opensource.org/licenses/MIT>)

at your option.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.

[`godot-logger`]: https://crates.io/crates/godot-logger
[`godot_print!`]: https://docs.rs/gdnative/latest/gdnative/macro.godot_print.html
[godot-rust]: https://godot-rust.github.io
[`log`]: https://crates.io/crates/log


CHANGELOG.md
<!-- markdownlint-disable-file MD024 -->

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2022-12-16

### Changed

- Relax version requirements for dependencies ([#39](https://github.com/jdno/godot-logger/pull/39))

## [1.0.0] - 2021-10-20

### Removed

- Remove `Filter` from the crate's public interface ([#13](https://github.com/jdno/godot-logger/pull/13))

## [0.3.0] - 2021-10-19

### Changed

- Replace `init` function with a builder pattern ([#11](https://github.com/jdno/godot-logger/pull/11))

## [0.2.0] - 2021-10-04

### Added

- Include module name in log output ([#7](https://github.com/jdno/godot-logger/pull/7))

## [0.1.0] - 2021-09-25

### Added

- Implement simple logger that prints messages to the output console in Godot

[unreleased]: https://github.com/jdno/godot-logger/compare/v1.1.0...HEAD
[1.1.0]: https://github.com/jdno/godot-logger/releases/tag/v1.1.0
[1.0.0]: https://github.com/jdno/godot-logger/releases/tag/v1.0.0
[0.3.0]: https://github.com/jdno/godot-logger/releases/tag/v0.3.0
[0.2.0]: https://github.com/jdno/godot-logger/releases/tag/v0.2.0
[0.1.0]: https://github.com/jdno/godot-logger/releases/tag/v0.1.0


src/filter.rs
use log::LevelFilter;

/// A filter applies a custom log level to a Rust module
///
/// Logs in [godot-logger] can be filtered using the default log level or a module-level override.
/// Module-level overrides are configured using a `Filter`, which combines a module path in Rust
/// with a log level.
///
/// [godot-logger]: https://crates.io/crates/godot-logger
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
pub(crate) struct Filter {
    module: &'static str,
    level: LevelFilter,
}

impl Filter {
    /// Initializes a new `Filter`
    ///
    /// Filters combine a module path in Rust with a log level, and are used to set a log level for
    /// a specific module.
    pub fn new(module: &'static str, level: LevelFilter) -> Self {
        Self { module, level }
    }

    /// Returns the filter's module path
    pub fn module(&self) -> &'static str {
        self.module
    }

    /// Returns the filter's log level
    pub fn level(&self) -> LevelFilter {
        self.level
    }
}

#[cfg(test)]
mod tests {
    use super::Filter;
    use log::LevelFilter;

    #[test]
    fn module() {
        let filter = Filter::new("godot_logger::filter", LevelFilter::Debug);
        assert_eq!(filter.module(), "godot_logger::filter");
    }

    #[test]
    fn level() {
        let filter = Filter::new("godot_logger::filter", LevelFilter::Debug);
        assert!(matches!(filter.level(), LevelFilter::Debug));
    }

    #[test]
    fn trait_send() {
        fn assert_send<T: Send>() {}
        assert_send::<Filter>();
    }

    #[test]
    fn trait_sync() {
        fn assert_sync<T: Sync>() {}
        assert_sync::<Filter>();
    }
}


src/builder.rs
use log::{Level, LevelFilter, SetLoggerError};
use log4rs::config::{Appender, Logger, Root};
use log4rs::Config;

use crate::appender::GodotAppender;
use crate::filter::Filter;

const APPENDER_NAME: &str = "godot-logger";

/// A `Builder` that configures and initializes the Godot logger
///
/// [godot-logger] implements the builder pattern as the primary interface to configure and
/// initialize the logger. The configuration has sensible defaults that can be overwritten by
/// calling the corresponding setters on the `Builder` struct. Once the configuration is done, the
/// logger can be initialized by calling the `build` method.
///
/// # Examples
///
/// ```
/// use log::Level;
/// use godot_logger::GodotLogger;
///
/// GodotLogger::builder()
///     .default_log_level(Level::Debug)
///     .init();
/// ```
///
/// [godot-logger]: https://crates.io/crates/godot-logger
#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
pub struct Builder {
    default_log_level: Level,
    filters: Vec<Filter>,
}

impl Builder {
    /// Sets the default log level
    ///
    /// `GodotLogger` matches all log records against a default log level. By default, only warnings
    /// and errors are logged.
    ///
    /// # Examples
    ///
    /// ```
    /// use log::Level;
    /// use godot_logger::GodotLogger;
    ///
    /// let mut builder = GodotLogger::builder();
    /// builder = builder.default_log_level(Level::Debug);
    /// ```
    pub fn default_log_level(mut self, default_log_level: Level) -> Self {
        self.default_log_level = default_log_level;
        self
    }

    /// Adds a filter
    ///
    /// Filters override the default log level for specific Rust modules.
    ///
    /// # Examples
    ///
    /// ```
    /// use godot_logger::GodotLogger;
    /// use log::LevelFilter;
    ///
    /// GodotLogger::builder().add_filter("godot_logger", LevelFilter::Off);
    /// ```
    pub fn add_filter(mut self, module: &'static str, level: LevelFilter) -> Self {
        self.filters.push(Filter::new(module, level));
        self
    }

    /// Initializes the logger
    ///
    /// This method consumes the builder and initializes the logger with the current configuration
    /// of the builder. After calling this method, log records will be written to Godot's output
    /// console.
    ///
    /// # Examples
    ///
    /// ```
    /// use log::Level;
    /// use godot_logger::GodotLogger;
    ///
    /// GodotLogger::builder().init();
    /// ```
    pub fn init(self) -> Result<(), SetLoggerError> {
        let loggers: Vec<Logger> = self
            .filters
            .iter()
            .map(|filter| {
                Logger::builder()
                    .appender(APPENDER_NAME)
                    .build(filter.module(), filter.level())
            })
            .collect();

        let config = Config::builder()
            .appender(Appender::builder().build(APPENDER_NAME, Box::new(GodotAppender)))
            .loggers(loggers)
            .build(
                Root::builder()
                    .appender(APPENDER_NAME)
                    .build(self.default_log_level.to_level_filter()),
            )
            .unwrap();

        let _handle = log4rs::init_config(config)?;
        Ok(())
    }
}

impl Default for Builder {
    fn default() -> Self {
        Self {
            default_log_level: Level::Warn,
            filters: Vec::new(),
        }
    }
}

#[cfg(test)]
mod tests {
    use log::{Level, LevelFilter};

    use super::Builder;

    #[test]
    fn default_log_level() {
        let mut builder = Builder::default();

        builder = builder.default_log_level(Level::Debug);

        assert!(matches!(builder.default_log_level, Level::Debug));
    }

    #[test]
    fn add_filter() {
        let mut builder = Builder::default();

        builder = builder.add_filter("godot_logger::builder", LevelFilter::Off);

        assert_eq!(builder.filters.len(), 1);
    }

    #[test]
    fn trait_default() {
        let builder = Builder::default();
        assert!(matches!(builder.default_log_level, Level::Warn));
    }

    #[test]
    fn trait_send() {
        fn assert_send<T: Send>() {}
        assert_send::<Builder>();
    }

    #[test]
    fn trait_sync() {
        fn assert_sync<T: Sync>() {}
        assert_sync::<Builder>();
    }
}


src/appender.rs
use chrono::Local;
use gdnative_core::{godot_print, godot_warn};
use log::{Level, Record};
use log4rs::append::Append;

/// A log appender for `log4rs` that prints to Godot's output console
///
/// [godot-logger] uses [log4rs] under the hood to configure the logger. A custom appender has been
/// created to write log records to Godot's output console.
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
pub(crate) struct GodotAppender;

impl Append for GodotAppender {
    fn append(&self, record: &Record) -> anyhow::Result<()> {
        let timestamp = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
        let level = record.level();
        let message = record.args();

        let output = match record.module_path() {
            Some(module) => format!("{} {} {} {}", timestamp, level, module, message),
            None => format!("{} {} {}", timestamp, level, message),
        };

        if record.level() <= Level::Warn {
            godot_warn!("{}", output);
        } else {
            godot_print!("{}", output);
        }

        Ok(())
    }

    fn flush(&self) {}
}


src/lib.rs
//! A simple logger that prints to Godot's output window
//!
//! [godot-logger] is a simple logger that prints log messages to the output console inside the
//! [Godot] game engine. It is built around the logging facade of the [log] crate, and uses the
//! [`godot_print!`] macro from the [gdnative] bindings.
//!
//! It is possible to configure different log levels for different Rust modules, similar to other
//! popular logging frameworks such as [env_logger] or [log4rs]. Simply provide a list as the second
//! argument to the `init` function with module names and log levels.
//!
//! # Use
//!
//! Add [godot-logger] and [log] as dependencies to `Cargo.toml`.
//!
//! Then initialize [godot-logger] in the `init` function that is exported by `gdnative`. Pass in a
//! default log level, and a list with module-level overrides (can be empty).
//!
//! ```
//! use gdnative::prelude::*;
//! use godot_logger::GodotLogger;
//! use log::{Level, LevelFilter};
//!
//! fn init(handle: InitHandle) {
//!     GodotLogger::builder()
//!         .default_log_level(Level::Info)
//!         .add_filter("godot_logger", LevelFilter::Debug)
//!         .init();
//!     log::debug!("Initialized the logger");
//! }
//!
//! godot_init!(init);
//! ```
//!
//! The following will appear in the _Output_ console inside Godot:
//!
//! ```text
//! 2021-09-25 19:29:25 DEBUG godot_logger Initialized the logger
//! ```
//!
//! [env_logger]: https://crates.io/crates/env_logger
//! [gdnative]: https://crates.io/crates/gdnative
//! [godot-logger]: https://crates.io/crates/godot-logger
//! [`godot_print!`]: https://docs.rs/gdnative/latest/gdnative/macro.godot_print.html
//! [log]: https://crates.io/crates/log
//! [log4rs]: https://crates.io/crates/log4rs
//! [Godot]: https://godotengine.org/

pub use crate::builder::*;

mod appender;
mod builder;
mod filter;

/// A logger that prints to the output console of the Godot game engine
///
/// `GodotLogger` is a logger implementation that prints log records to the output console inside
/// the Godot game engine. The log level can be set per Rust module, similar to other logging
/// frameworks in Rust.
///
/// The recommended way to initialize the logger is by using the crate's [`Builder`]. Its setters
/// can be used to configure the logger and overwrite the default configuration.
///
/// # Examples
///
/// ```
/// use godot_logger::GodotLogger;
/// use log::{Level, LevelFilter};
///
/// // Configure and initialize the logger
/// GodotLogger::builder()
///     .default_log_level(Level::Debug)
///     .add_filter("godot-logger", LevelFilter::Warn)
///     .init();
/// ```
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
pub struct GodotLogger;

impl GodotLogger {
    pub fn builder() -> Builder {
        Builder::default()
    }
}
